### PR TITLE
fix gcs::RedisAsioClient non-thread safe

### PR DESCRIPTION
**Pull request by micafan**
## Why are these changes needed?
We might call gcs::RedisAsioClient in multi-threads(e.g. c++ core worker), but [boost socket is non-thread safe](https://www.boost.org/doc/libs/1_68_0/doc/html/boost_asio/reference/ip__tcp/socket.html), so we dispatch all socket operations to a single thread(backend thread), to make this client thread safe.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/gcs/asio.cc [hunk 1]
```diff
@@ -67,13 +68,23 @@ void RedisAsioClient::handle_write(boost::system::error_code error_code) {
...
 }
 
 void RedisAsioClient::del_read() { read_requested_ = false; }
 
 void RedisAsioClient::add_write() {
+  // Socket is non-thread safe, dispatch write to backend thread.
+  io_service_.dispatch(boost::bind(&RedisAsioClient::do_add_write, this));
+}
+
+void RedisAsioClient::do_add_write() {
```
Comment (edoakes) [inline src/ray/gcs/asio.cc]: Can we just make this a lambda in `add_write` for readability?

**Review by edoakes — APPROVED** [review]
LGTM

Could you also add a disclaimer comment in the header of the `RedisAsioClient` and above the core worker `gcs_client_->Connect()` call that specifies that this needs to run on a single-threaded `io_service`?

Also, it seem that we can entirely remove the logic in `redis_async_context.cc` after this because that was only added to make the client thread safe - right?

File: src/ray/gcs/asio.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/gcs/asio.cc]: same for `do_add_read` below

Comment (micafan) [reply src/ray/gcs/asio.cc]: Good point.

**Comment by micafan** [comment]
> LGTM
> 
> Could you also add a disclaimer comment in the header of the `RedisAsioClient` and above the core worker `gcs_client_->Connect()` call that specifies that this needs to run on a single-threaded `io_service`?

Thanks for the suggestion. Already add comments to RedisGcsClient and RedisAsioClient. 

> Also, it seem that we can entirely remove the logic in `redis_async_context.cc` after this because that was only added to make the client thread safe - right?

We can't remove the logic in `redis_async_context.cc` yet.  

Redis write commands may run on both background thread and user threads.
User may initiate gcs requests (e.g.: RedisContext::RunAsync) from multiple threads and trigger redis wirte commands. These commands are not dispatched to the background thread, but run on the user threads. So we need lock to ensure the execution of the commands are thread safe.

Redis read commands will only run on background threads.
Redis read commands are always triggered by Redis write comands, and RedisAsioClient dispatches these redis read commands to the background thread, so there is no need to lock.

**Comment by edoakes** [comment]
Ran the GCS tests that were failing in a loop on travis:
https://github.com/ray-project/ray/pull/6199/checks?check_run_id=310856034

Unfortunately it looks like we still get the same error as before:
`F1120 01:32:41.198681  4044 redis_context.cc:147]  Check failed: callback_items_.find(callback_index) != callback_items_.end()`

However, it is substantially improved as the tests ran for much longer and we didn't get this error every time. @micafan could you take a look at why we're still getting that assertion? In the meantime, merging this in the hopes that it fixes the user-facing error reported in https://github.com/ray-project/ray/issues/6169.

**Comment by micafan** [comment]
hi, @edoakes , I already found the root cause of this bug.  It’s my bad. I did’t noticed that the callback  list of hiredis async context are non-thread safe. The details is:

Both r`edisvAsyncCommand(…)` and `redisAsyncHandleRead(…)` will change the callback list of `asyncRedisContext`.
The diffrent is `redisvAsyncCommand` adds new callback to `asyncRedisContext's` callback list while `redisAsyncHandleRead` consumes callback from it.
The call stack is：
    `redisvAsyncCommand -> __redisAsyncCommand-> __redisPushCallback`
and
    `redisAsyncHandleRead -> redisProcessCallbacks-> __redisShiftCallback `

In class `RedisAsyncContext`, `redisvAsyncCommand` is locked, but `redisAsyncHandleRead` isn’t. There may be two exceptions,  one is the consumed callback  isn’t remove from callback list, and the other is the callback isn’t added to callback list. 

I should use a lock to protect the callback list like you did in [PR6145](https://github.com/ray-project/ray/pull/6145). And to avoid deadlock that we discussed before, we should make sure the callback always executed outside of the lock, e.g. by dispatch the callback to the event loop thread.

**Comment by micafan** [comment]
Once the callback becomes asynchronous, a copy of the data is inevitable and adds complexity to the code.

**Comment by micafan** [comment]
[Probably not the best solution](https://github.com/ray-project/ray/pull/6222), open to discuss.

**Comment by edoakes** [comment]
I have your suggested fix working in the following PR and verified that it works. Let's move discussion there please.

https://github.com/ray-project/ray/pull/6231

**Comment by micafan** [comment]
> I have your suggested fix working in the following PR and verified that it works. Let's move discussion there please.
> 
> #6231

Sounds good!

**Pull request merged**